src/stdlib_linalg_expand.fypp

```diff
@@ -0,0 +1,74 @@
+#:include "common.fypp"
+#:set ALL_KINDS_TYPES = REAL_KINDS_TYPES + INT_KINDS_TYPES + CMPLX_KINDS_TYPES &
+    & + LOG_KINDS_TYPES + STRING_KINDS_TYPES
+submodule(stdlib_linalg) stdlib_linalg_expand
+
+contains
+
+    !> `ones` creates a vector, filled completely with `1` `integer` type values.
+    pure module function ones_1_default(dim) result(result)
+        integer, intent(in) :: dim
+        integer, allocatable :: result(:)
+
+        allocate(result(dim))
+        result = 1
+
+    end function ones_1_default
+
+    !> `ones` creates a matrix, filled completely with `1` `integer` type values.
+    pure module function ones_2_default(dim1, dim2) result(result)
+        integer, intent(in) :: dim1, dim2
+        integer, allocatable :: result(:, :)
+
+        allocate(result(dim1, dim2))
+        result = 1
+
+    end function ones_2_default
+
+    !> `zeros` creates a vector, filled completely with `0` `integer` type values.
+    pure module function zeros_1_default(dim) result(result)
+        integer, intent(in) :: dim
+        integer, allocatable :: result(:)
+
+        allocate(result(dim))
+        result = 0
+
+    end function zeros_1_default
+
+    !> `zeros` creates a matrix, filled completely with `0` `integer` type values.
+    pure module function zeros_2_default(dim1, dim2) result(result)
+        integer, intent(in) :: dim1, dim2
+        integer, allocatable :: result(:, :)
+
+        allocate(result(dim1, dim2))
+        result = 0
+
+    end function zeros_2_default
+
+    #:for k1, t1 in ALL_KINDS_TYPES
+    !> `expand` creates a vector, filled completely with `value` `${t1}$` type values.
+    pure module function expand_1_${t1[0]}$_${k1}$(value, dim) result(result)
+        
+        ${t1}$, intent(in) :: value
+        integer, intent(in) :: dim
+        ${t1}$, allocatable :: result(:)
+
+        allocate(result(dim))
+        result = value
+        
+    end function expand_1_${t1[0]}$_${k1}$
```

Comment:
I was thinking, is there a need for `expand` function?
I thought it existed to make up for the existing shortcomings of `zeros` and `ones`.

And it also manipulates `logical` and `string_type` arguments, which appear in `stdlib_linalg`, which seems out of place. (I would remove the support for `logical` and `string_type` arguments)

Reply:
I was rather thinking why it is named as `expand` when it is not expanding anything.

Reply:
I think if the support of `logical` and `string_type` is removed. The functions of `expand` and `zeros/ones` overlap too much, and the difference is not too big. I may consider removing the `expand` function, focusing on the fair use  of the current stage  `zeros/ones` (experimental version).

Reply:
> I was rather thinking why it is named as `expand` when it is not expanding anything.

Haha😜, It expands the `value` value to a vector or a matrix. (the API name is discussed here, https://github.com/fortran-lang/stdlib/pull/478#discussion_r679893172)